rclpy/src/rclpy/action_server.cpp

```diff
@@ -0,0 +1,418 @@
+    throw rclpy::RCLError("Failed get goal status array");
+  }
+
+  ret = rcl_action_publish_status(rcl_action_server_.get(), &status_message);
+
+  if (RCL_RET_OK != ret) {
+    throw rclpy::RCLError("Failed publish goal status array");
+  }
+}
+
+void
+ActionServer::notify_goal_done()
+{
+  rcl_ret_t ret = rcl_action_notify_goal_done(rcl_action_server_.get());
+  if (RCL_RET_OK != ret) {
+    throw rclpy::RCLError("Failed to notfiy action server of goal done");
+  }
+}
+
+bool
+ActionServer::goal_exists(py::object pygoal_info)
+{
+  auto goal_info = convert_from_py(pygoal_info);
+  rcl_action_goal_info_t * goal_info_type = static_cast<rcl_action_goal_info_t *>(goal_info.get());
+  if (!goal_info_type) {
+    throw py::error_already_set();
+  }
+  return rcl_action_server_goal_exists(rcl_action_server_.get(), goal_info_type);
+}
+
+py::tuple
+ActionServer::get_num_entities()
+{
+  size_t num_subscriptions = 0u;
+  size_t num_guard_conditions = 0u;
+  size_t num_timers = 0u;
+  size_t num_clients = 0u;
+  size_t num_services = 0u;
+
+  rcl_ret_t ret = rcl_action_server_wait_set_get_num_entities(
+    rcl_action_server_.get(),
+    &num_subscriptions,
+    &num_guard_conditions,
+    &num_timers,
+    &num_clients,
+    &num_services);
+
+  if (RCL_RET_OK != ret) {
+    std::string error_text{"Failed to get number of entities for 'rcl_action_server_t'"};
+    throw rclpy::RCLError(error_text);
+  }
+
+  py::tuple result_tuple(5);
+  result_tuple[0] = py::int_(num_subscriptions);
+  result_tuple[1] = py::int_(num_guard_conditions);
+  result_tuple[2] = py::int_(num_timers);
+  result_tuple[3] = py::int_(num_clients);
+  result_tuple[4] = py::int_(num_services);
+  return result_tuple;
+}
+
+py::tuple
+ActionServer::is_ready(py::capsule pywait_set)
+{
+  auto wait_set = static_cast<rcl_wait_set_t *>(pywait_set);
+  bool is_goal_request_ready = false;
+  bool is_cancel_request_ready = false;
+  bool is_result_request_ready = false;
+  bool is_goal_expired = false;
+  rcl_ret_t ret = rcl_action_server_wait_set_get_entities_ready(
+    wait_set,
+    rcl_action_server_.get(),
+    &is_goal_request_ready,
+    &is_cancel_request_ready,
+    &is_result_request_ready,
+    &is_goal_expired);
+
+  if (RCL_RET_OK != ret) {
+    throw rclpy::RCLError("Failed to get number of ready entities for action server");
+  }
+
+  py::tuple result_tuple(4);
+  result_tuple[0] = py::bool_(is_goal_request_ready);
+  result_tuple[1] = py::bool_(is_cancel_request_ready);
+  result_tuple[2] = py::bool_(is_result_request_ready);
+  result_tuple[3] = py::bool_(is_goal_expired);
+  return result_tuple;
+}
+
+void
+ActionServer::add_to_waitset(py::capsule pywait_set)
+{
+  auto wait_set = static_cast<rcl_wait_set_t *>(pywait_set);
+  rcl_ret_t ret = rcl_action_wait_set_add_action_server(wait_set, rcl_action_server_.get(), NULL);
+  if (RCL_RET_OK != ret) {
+    std::string error_text{"Failed to add 'rcl_action_server_t' to wait set"};
+    throw rclpy::RCLError(error_text);
```

Comment:
@ahcorde nit: why not collapsing this two in one line? 
```suggestion
    throw rclpy::RCLError("Failed to add 'rcl_action_server_t' to wait set");
```